### PR TITLE
benchmark: Add MXFP4/MXFP8 quantization mode support to FP4 MoE benchmark

### PR DESCRIPTION
**Pull request by bkryu**
## 📌 Description

* Add --fp4_mode CLI argument to trtllm_fp4_block_scale_moe benchmark with three modes:
  * nvfp4 (default, existing behavior): NvFP4 weights + NvFP4 hidden states, block_size=16
  * mxfp4_bf16: MXFP4 weights + BF16 hidden states, block_size=32
  * mxfp4_mxfp8: MXFP4 weights + MXFP8 hidden states, block_size=32
* Add backward compatibility with flashinfer 0.6.0 (pre-0.6.3), where ActivationType was not yet exported from the top-level package and MoE APIs used gated_act_type instead of activation_type
* Fix CSV output column mismatch: moe.py wrote to cur_res["activation_type"] but the CSV column was "gated_act", causing the field to be silently empty

## 🚀 Pull Request Checklist

Thank you for contributing to FlashInfer! Before we review your pull request, please make sure the following items are complete.

### ✅ Pre-commit Checks

- [x] I have installed `pre-commit` by running `pip install pre-commit` (or used your preferred method).
- [x] I have installed the hooks with `pre-commit install`.
- [x] I have run the hooks manually with `pre-commit run --all-files` and fixed any reported issues.

> If you are unsure about how to set up `pre-commit`, see [the pre-commit documentation](https://pre-commit.com/).

## 🧪 Tests

- [x] Tests have been added or updated as needed.
- [x] All tests are passing (`unittest`, etc.).

## Summary by CodeRabbit

* **New Features**
  * Added FP4 mode selection for benchmarks (nvfp4 default, mxfp4_mxfp8, mxfp4_bf16) with a CLI flag and updated benchmark output to include activation type and FP4 mode.
  * Benchmarks now report FP-format-aware size/scale behavior for new modes.

* **Chores**
  * Backward-compatible handling of older activation argument names.
  * Adjusted quantization and vector-size logic to support mxfp4/mxfp8 modes.

### CODE REVIEW

File: benchmarks/routines/moe.py [hunk 1]
```diff
@@ -496,52 +539,56 @@ def testTrtllmFp4BlockScaleMoe(args):
...
-    if hidden_states_scale_linear_fp4.numel() != expected_scale_elems:
-        if args.verbose >= 1:
-            print(
-                f"[INFO] Adjusting FP4 hidden_states_scale from {hidden_states_scale_linear_fp4.numel()} to {expected_scale_elems} elements"
+    # Prepare hidden states and scale based on fp4_mode
+    if fp4_mode == "mxfp4_bf16":
+        hidden_states_fp4 = hidden_states.to(torch.bfloat16)
+        hidden_states_scale_linear_fp4 = None
+    elif fp4_mode == "mxfp4_mxfp8":
+        hs_quant, hs_scale = mxfp8_quantize(hidden_states, True)
```
[the remark this answers is not in the document]
Comment (bkryu) [reply benchmarks/routines/moe.py:560]: The MoE kernel requires the swizzled layout to be True. Will raise a value error if not a multipleof 128.

File: benchmarks/routines/moe.py [hunk 2]
```diff
@@ -496,52 +539,56 @@ def testTrtllmFp4BlockScaleMoe(args):
...
+        hidden_states_fp4 = hs_quant
+        hidden_states_scale_linear_fp4 = hs_scale.view(torch.float8_e4m3fn).reshape(
+            num_tokens, -1
+        )
+    else:
+        hidden_states_scale_global = calculate_fp4_global_scale(hidden_states)
+        hidden_states_fp4_bytes, hidden_states_scale_fp4_bytes, _ = quantize_fp4(
+            hidden_states, hidden_states_scale_global, use_ue8m0, True
+        )
```
[the remark this answers is not in the document]
Comment (bkryu) [reply benchmarks/routines/moe.py:569]: The MoE kernel expects a swizzled layout. This must be true.

File: benchmarks/routines/moe.py [hunk 3]
```diff
@@ -496,52 +539,56 @@ def testTrtllmFp4BlockScaleMoe(args):
...
+    # Prepare hidden states and scale based on fp4_mode
+    if fp4_mode == "mxfp4_bf16":
+        hidden_states_fp4 = hidden_states.to(torch.bfloat16)
+        hidden_states_scale_linear_fp4 = None
+    elif fp4_mode == "mxfp4_mxfp8":
+        hs_quant, hs_scale = mxfp8_quantize(hidden_states, True)
+        hidden_states_fp4 = hs_quant
+        hidden_states_scale_linear_fp4 = hs_scale.view(torch.float8_e4m3fn).reshape(
+            num_tokens, -1
+        )
```
[the remark this answers is not in the document]
Comment (bkryu) [reply benchmarks/routines/moe.py:564]: Not true. 

What's happening here is that mxfp8_quantize returns scale factors as torch.uint8 because PyTorch has no native ue8m0 dtype. The .view(torch.float8_e4m3fn) is not a semantic conversion -- it's using float8_e4m3fn as a carrier dtype (both are 1 byte per element). The C++ GEMM kernel knows the actual bit layout is UE8M0 and interprets the bytes correctly regardless of what PyTorch dtype label is on the tensor. This is a standard pattern throughout the flashinfer quantization code (you'll see the same .view(torch.float8_e4m3fn) for NvFP4 scales too).
No change needed here.

File: benchmarks/routines/moe.py [hunk 4]
```diff
@@ -43,6 +50,23 @@
...
+}
+
+
+def _activation_kwarg(fn, activation_type: ActivationType) -> dict:
+    """Return the correct activation keyword argument for *fn* in the installed version."""
+    sig = inspect.signature(fn)
+    if "activation_type" in sig.parameters:
+        return {"activation_type": activation_type.value}
+    if "gated_act_type" in sig.parameters:
+        return {"gated_act_type": _ACTIVATION_TO_GATED_ACT.get(activation_type, 0)}
```
[the remark this answers is not in the document]
Comment (bkryu) [reply benchmarks/routines/moe.py]: Will address in subsequent commit

File: benchmarks/routines/moe.py [hunk 5]
```diff
@@ -496,52 +539,56 @@ def testTrtllmFp4BlockScaleMoe(args):
...
+        expected_scale_elems = (num_tokens * hidden_size) // sf_vec_size
+        if hidden_states_scale_linear_fp4.numel() != expected_scale_elems:
+            if args.verbose >= 1:
+                print(
+                    f"[INFO] Adjusting FP4 hidden_states_scale from "
+                    f"{hidden_states_scale_linear_fp4.numel()} to {expected_scale_elems} elements"
+                )
+            hidden_states_scale_linear_fp4 = torch.ones(
+                expected_scale_elems, device=device, dtype=torch.float8_e4m3fn
             )
```
[the remark this answers is not in the document]
Comment (bkryu) [reply benchmarks/routines/moe.py:586]: Addressed in latest commit

File: benchmarks/routines/moe.py [hunk 6]
```diff
@@ -187,6 +217,20 @@ def parse_moe_args(line, parser):
...
     )
+    parser.add_argument(
+        "--fp4_mode",
+        type=str,
+        required=False,
+        default="nvfp4",
+        choices=["nvfp4", "mxfp4_mxfp8", "mxfp4_bf16"],
+        help=(
+            "FP4 quantization mode for trtllm_fp4_block_scale_moe: "
+            "nvfp4 (NvFP4 weights + NvFP4 hidden states, block_size=16), "
```
Comment (yzh119) [inline benchmarks/routines/moe.py:229]: If these block_size are hardcoded numbers, we can create a utility function or a dict to store the block_size, e.g.:
```
sf_vec_size = {
    "nvfp4": 16,
    "mxfp4": 32,
}
```

File: benchmarks/routines/moe_utils.py [hunk 7]
```diff
@@ -542,6 +542,9 @@ def get_effective_bytes(
         elif fmt == "mxfp4":
             # 1 e2m1 + 1 ue8m0 scale per 32-element block
             return 0.5 + 1 / 32
+        elif fmt == "mxfp8":
+            # 1 e4m3 + 1 ue8m0 scale per 32-element block
+            return 1.0 + 1 / 32
```
Comment (yzh119) [inline benchmarks/routines/moe_utils.py]: ditto

File: benchmarks/routines/moe.py (hunk 6 above)
Comment (bkryu) [reply benchmarks/routines/moe.py:229]: Addressed in latest commit [d149911](https://github.com/flashinfer-ai/flashinfer/pull/2635/changes/d14991150036bf4145d4bff9b22ed5bea0700a1a)

File: benchmarks/routines/moe_utils.py (hunk 7 above)
Comment (bkryu) [reply benchmarks/routines/moe_utils.py]: Ditto; addressed in latest commit [d149911](https://github.com/flashinfer-ai/flashinfer/pull/2635/changes/d14991150036bf4145d4bff9b22ed5bea0700a1a)
